crates/revm/src/evm_impl.rs

```diff
@@ -161,7 +164,71 @@ impl<'a, GSPEC: Spec, DB: Database> EVMImpl<'a, GSPEC, DB> {
 }
 
 impl<'a, GSPEC: Spec + 'static, DB: Database> Transact<DB::Error> for EVMImpl<'a, GSPEC, DB> {
+    #[inline]
     fn preverify_transaction(&mut self) -> Result<(), EVMError<DB::Error>> {
+        self.preverify_transaction_inner()
+    }
+
+    #[inline]
+    fn transact_preverified(&mut self) -> EVMResult<DB::Error> {
+        let output = self.transact_preverified_inner();
+        self.handler.end(&mut self.data, output)
+    }
+
+    #[inline]
+    fn transact(&mut self) -> EVMResult<DB::Error> {
+        let output = self
+            .preverify_transaction()
+            .and_then(|()| self.transact_preverified());
+        self.handler.end(&mut self.data, output)
+    }
```

Comment:
This looks safe since OP's end handler will convert the `Transaction` error over to a `Halt` variant, making the double-run not happen, but is a bit confusing at first glance. Is there a reason we've split up the end handler calls like this rather than just chaining in `transact`?

Reply:
Yeah, this is not correct, and we should use `*_inner()` calls. Will make the change.